### ISSUE TITLE
Understanding the implementation w.r.t. the paper

### ISSUE THREAD

**Issue by anandsaha** [loshchil/AdamW-and-SGDW#1]
Hello,

Thanks for this paper!

I am trying to understand the paper and it's implementation here and have some doubts. There are a couple of places where the formulas in the paper doesn't match the implementation. I am sure I am missing some link, so can you please shed some light here:

1. The normalized decay equation (5) page 4, and it's implementation at line 76 of train.lua: I am not able to directly map the paper's formula to the implementation. 
2. In the paper we have a SetScheduleMultiplier(t) in Algorithm 2, which is not in the paper: Are we setting eta as a fixed value of 1?
3. Also the way we are adding the regularization term at line 12 of Algorithm 2 is not directly relatable to the code.

Am I missing any prerequisite here? Not that I haven't read the shake-shake paper and trying to understand only this one.

Thanks!
Anand

**Comment by loshchil** [comment]
Hi Anand,

1. In line 76 of train.lua: trainSize is the number of batches per epoch and self.opt.Te is the number of epochs to be performed for that run / restart. Thus, we get the total number of batch passes in this run / restart which we set to power of 0.5, i.e., sqrt, to obtain the actual weight decay factor from its normalized value of self.optimState.weightDecay.

2. Please have a look at learningRateMultiplier which is controlled by learningRateCosineSGDR(). The same eta multiplier is used for weightDecayMultiplier.

3. The answer is related to the one given before. Note that weightDecaycurrent represents weightDecayMultiplier * self.optimState.weightDecay.  

Please let me know if my answers clarify your questions. 
Note that the github code is a cleaned up version (still not really clean) of the code that I used for my experiments. I hope I didn't break / clean something important.

Thank for your interest,
Ilya

**Comment by anandsaha** [comment]
Hello Ilya,

That was very helpful and makes things more clear to me!

May I ask you a couple of followup question:

train.lua line 76:
```
local weightDecayNormalized = self.optimState.weightDecay / ( math.pow(trainSize * self.opt.Te, 0.5) )
```

Q1. This is a very crucial line since this decays the weight. Here I do not see the batch size being taken into account, unlike formula (5) in the paper.

Q2. I also do not see learningRateMultiplier (eta) being applied to the entire contents of the bracket, as seen in line 12 of Algorithm 2. It is being applied only to the L2 regularization part in the code.

Thanks!
Anand

**Comment by anandsaha** [comment]
May I also add: You have applied your approach on top of Shake-Shake regularizer. Would this be your recommendation for adopting AdamWR? Do you have any benchmarks for AdamWR alone?

Thanks!
Anand

**Comment by loshchil** [comment]
> Q1. This is a very crucial line since this decays the weight. Here I do not see the batch size being taken into account, unlike formula (5) in the paper.

As I mentioned in my previous reply, trainSize is not the number of training points but the number of batches (total number of training examples / batch size) and thus takes into account batch size.

> Q2. I also do not see learningRateMultiplier (eta) being applied to the entire contents of the bracket, as seen in line 12 of Algorithm 2. It is being applied only to the L2 regularization part in the code.

eta sets learningRateMultiplier and weightDecayMultiplier and eta * (A + B) is the same as eta * a + eta * B, i.e., learningRateMultiplier * A + weightDecayMultiplier * B.

> May I also add: You have applied your approach on top of Shake-Shake regularizer. Would this be your recommendation for adopting AdamWR? Do you have any benchmarks for AdamWR alone?

In order to use "AdamWR alone", you can switch-off the Shake-Shake part by having a few "False" settings in the input command line - see the original repo of Shake-Shake. AdamWR is independent in its design on Shake-Shake and you don't have to use Shake-Shake at all. The only reason Shake-Shake is used here is in order to benchmark in the regime of SOTA results.

**Comment by anandsaha** [comment]
Hello Ilya,

Thanks for patiently answering my questions!

I am happy to share that I have been able to replicate your results. [Here](https://github.com/anandsaha/fastai.part1.v2/blob/master/courses/dl1/adam-experiment-cifar10.ipynb) is a notebook where I got better results with AdamW and SGDW approach in PyTorch (scroll to the last graph at the end)

I am a [fast.ai](http://fast.ai) student and @jph00 ( [Jeremy Howard](https://twitter.com/jeremyphoward) ) had pointed us to your paper and motivated me to try to implement it in [fastai library](https://github.com/fastai/fastai). 

I appreciate your work. In the process, I learned a lot.

Today I noticed that there are couple of PR's to take your work  [SGDR](https://github.com/pytorch/pytorch/pull/3311) and [AdamW/SGDW](https://github.com/pytorch/pytorch/issues/3790) to PyTorch. So, very soon we will get it out of the box :) 

**Thanks!**

--

**Comment by loshchil** [comment]
Hi Anand @anandsaha ,

Thanks for your contribution and notebook! I am glad to find that the results look very good.
That should help me to finally move from Torch to PyTorch :-) 

All the best,
Ilya

**Comment by jph00** [comment]
I actually think that we can use this approach to significantly simplify how pytorch handles weight decay. Since it happens outside of the grad and grad^2 calculations, I plan to move regularization out of the optimizer entirely and into the training loop - which is where I think it really belongs.

So I think your paper @loshchil can help simplify code, as well as making training faster!

**Comment by loshchil** [comment]
Thanks for your interest in my work, @jph00 !

**Comment by jph00** [comment]
@loshchil I'm more than interested - I'm a big fan :) Not just of this, but of all the papers you and Hutter have done.

Anyway - you might be interested in checking out the implementation that @anandsaha has now completed. You'll see callbacks used in a clever way to separate out the concerns of the weight decay class vs the optim class, as I hope in my earlier comment: https://github.com/fastai/fastai/pull/46/files

**Comment by loshchil** [comment]
@jph00 it is flattering to know that you like our work! 

I shall have a look at the implementation, hopefully soon when I will move with my research trials to PyTorch and get more familiar with it. Soon we will know what ICLR reviewers demand us to modify, 2 years ago the reviewers booked me for 2 months :-)
